File: Trie/139.Word-Break/139.Word-Break.java
class Solution {
    private Node root;
    private int[] memo;
    private class Node {
        private boolean isKey;
        private HashMap<Character, Node> next;
        private Node(boolean isKey) {
            this.isKey = isKey;
            next = new HashMap<>();
        }
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        initTrie(wordDict);
        memo = new int[s.length()];
        return dfs(0, s.length(), s);
    }
    private void initTrie(List<String> wordDict) {
        root = new Node(false);
        Node curr = root;
        for (String word : wordDict) {
            for (char c: word.toCharArray()) {
                if (!curr.next.containsKey(c)) {
                    curr.next.put(c, new Node(false));
                }
                curr = curr.next.get(c);
            }
            curr.isKey = true;
            curr = root;
        }
    }
    private boolean dfs(int index,int sLen, String s) {
        if (index == sLen) {
            return true;
        }

        if (memo[index] == 1) {
            return false;
        }

        Node curr = root;
        for (int i = index; i < sLen; i++) {
            char c = s.charAt(i);
            if (curr.next.containsKey(c)) {
                curr = curr.next.get(c);
                if (curr.isKey && dfs(i + 1, sLen, s)) {
                    return true;
                }
            } else {
                break;
            }
        }
        memo[index] = 1;
        return false;
    }
}

File: Trie/139.Word-Break/139.Word-Break.ts
class TrieNode {
  public next: Array<TrieNode | null>;
  public isEnd: Boolean;
  public constructor() {
    this.next = new Array(26);
    this.isEnd = false;
    for (let i = 0; i < 26; i++) {
      this.next[i] = null;
    }
  }
}

function wordBreak(s: string, wordDict: string[]): boolean {
  const root = new TrieNode();
  const memo = new Array(301).fill(0);

  for (const word of wordDict) {
    let node = root;
    for (const char of word) {
      if (node.next[char.charCodeAt(0) - 97] === null) {
        node.next[char.charCodeAt(0) - 97] = new TrieNode();
      }
      node = node.next[char.charCodeAt(0) - 97];
    }
    node.isEnd = true;
  }

  return dfs(s, 0, root, memo);
}

function dfs(
  s: string,
  index: number,
  root: TrieNode,
  memo: number[]
): boolean {
  if (index === s.length) {
    return true;
  }

  if (memo[index] === 1) {
    return false;
  }

  let node = root;
  for (let i = index; i < s.length; i++) {
    if (node.next[s[i].charCodeAt(0) - 97] !== null) {
      node = node.next[s[i].charCodeAt(0) - 97];
      if (node.isEnd && dfs(s, i + 1, root, memo)) {
        return true;
      }
    } else {
      break;
    }
  }
  memo[index] = 1;
  return false;
}
